### DISCUSSION TITLE
Does GitOps apply to on-prem infrastructure?

### DISCUSSION THREAD

**Discussion by aklyachkin** [open-gitops/project#101] [Q&A]
GitOps is built upon [4 principles](https://github.com/open-gitops/documents/blob/main/PRINCIPLES.md). Using configuration management tools such Ansible, Chef, Puppet, and so on, and saving the desired configuration in a Git repository, it would be possible to achieve it:

- Declarative language is in use by all named configuration management tools
- If we save our playbooks/cookbooks/manifests in a Git repo (which is best practice for all tools), you have immutability, versioning, and a complete version history
- Client-oriented tools like chef pull the desired state declarations automatically. For Ansible it can be achieved by using Tower/AWX.
- They attempt to apply the desired state, as it is described in the declaration.

So far so good. The only missing parts I see:

- saving in a Git repo is best practice for the tools, but not the requirement
- some of the tools (like Ansible) are not pull-based, but push-based

@chris-short brought in, that Kubernetes API gives a capability to declare the entire state of the infrastructure and application "in one feel swoop". It is a big difference between config mgmt approach, when you configure separate assets like servers, and Kubernetes approach, when you may configure the whole cluster with all applications on it.

OTOH you first need an installed Kubernetes cluster and we come back to bare-metal installations and how to describe them in GitOps?

But the main question I have is in the title. Does GitOps apply to on-prem or private IAAS Cloud (non-Kubernetes) infrastructure? If yes, how can it be applied? Any blueprints?

**Comment by chris-short** [comment]
This conversation started on Twitter: https://twitter.com/aklyachkin/status/1523698528299761666

I welcome this conversation as do the other co-chairs, I'm sure.

**Comment by christianh814** [comment]
> GitOps is built upon [4 principles](https://github.com/open-gitops/documents/blob/main/PRINCIPLES.md). Using configuration management tools such Ansible, Chef, Puppet, and so on, 

No. Using these tools + Git **IS NOT** GitOps. More specifically these tools, themselves, aren't GitOps because of[ Principle #4](https://github.com/open-gitops/documents/blob/release-v1.0.0/PRINCIPLES.md). 

Example: 

Terraform + Git, **is not** GitOps

Terraform Kubernetes Operator + Git **is** GitOps

It's not that these tools are "bad", it's that they're missing the continuous reconciliation. 

> and saving the desired configuration in a Git repository, it would be possible to achieve it:
> 
>     * Declarative language is in use by all named configuration management tools
> 
>     * If we save our playbooks/cookbooks/manifests in a Git repo (which is best practice for all tools), you have immutability, versioning, and a complete version history
> 
>     * Client-oriented tools like chef pull the desired state declarations automatically. For Ansible it can be achieved by using Tower/AWX.
> 
>     * They attempt to apply the desired state, as it is described in the declaration.
> 
> 
> So far so good. The only missing parts I see:
> 
>     * saving in a Git repo is best practice for the tools, but not the requirement
> 
>     * some of the tools (like Ansible) are not pull-based, but push-based

Don't get this confused. Principle #3 doesn't mean pull vs push. It means the Git contents are pulled. But **HOW** it gets applied, doesn't matter. 

More clearly: The GitOps Controller retrieves the manifests from git, then the GitOps Controller can push those configurations to remote clusters.

The idea behind Principle #3 is to differentiate it from Webhooks (i.e. github actions). 

I'm only saying this to make it clear. How the manifests end up on the target system from the GitOps controller doesn't necessarily matter, as long as the manifests (from Git or from s3) is PULLED by the GitOps controller. 

> 
> 
> @chris-short brought in, that Kubernetes API gives a capability to declare the entire state of the infrastructure and application "in one feel swoop". It is a big difference between config mgmt approach, when you configure separate assets like servers, and Kubernetes approach, when you may configure the whole cluster with all applications on it.
> 
> OTOH you first need an installed Kubernetes cluster and we come back to bare-metal installations and how to describe them in GitOps?
> 
> But the main question I have is in the title. Does GitOps apply to on-prem or private IAAS Cloud (non-Kubernetes) infrastructure? If yes, how can it be applied? Any blueprints?

**Reply by aklyachkin** [reply]
Thank you Christian. Let me to clarify some points.

> Example:
> Terraform + Git, is not GitOps

I fully agree with you in case with Terraform. It doesn't have any capabilities to reconcile infrastructure per se and if we want to have Terraform in a GitOps environment, we should have some "reconciliation agent", such as CI/CD.

But the tools like Chef and Puppet do have such agents. They do exactly what is stated in Principle 4:
> Software agents [continuously](https://github.com/open-gitops/documents/blob/release-v1.0.0/GLOSSARY.md#continuous) observe actual system state and [attempt to apply](https://github.com/open-gitops/documents/blob/release-v1.0.0/GLOSSARY.md#reconciliation) the desired state.

The difference I see, if we look into the [glossary](https://github.com/open-gitops/documents/blob/release-v1.0.0/GLOSSARY.md#reconciliation), is that it is timely triggered reconciliation (in case with Chef, Puppet or AWX schedule), but not drift-driven. 

For me it is very difficult to understand this difference. Let's take it by another example. My Chef agent runs approx. every 30 minutes, it downloads its cookbooks and checks if there are configuration drifts from the desired state. If there are, it takes the required actions to bring the system back to the desired state.

If I read the principle 4 and the glossary, what does it mean:

> "Continuous" is intended to match the industry standard term: [reconciliation](https://github.com/open-gitops/documents/blob/release-v1.0.0/GLOSSARY.md#reconciliation) continues to happen, not that it must be instantaneous.

OK, Chef agent tries to do it every 30 minutes, so it is continuous from my point of view.

> The process of ensuring the actual state of a system matches its [desired state](https://github.com/open-gitops/documents/blob/release-v1.0.0/GLOSSARY.md#desired-state).

Yes, I see here the same behaviour.

>  Contrary to traditional CI/CD where automation is generally driven by pre-set triggers, in GitOps reconciliation is triggered whenever there is a divergence.

In this case Chef agent is triggered by time, but it reconciles only if there is a divergence.

> Divergence could be due to the actual state unintentionally [drifting](https://github.com/open-gitops/documents/blob/release-v1.0.0/GLOSSARY.md#drift) from the desired state declarations, or a new desired state declaration version having been changed intentionally.

This is a reason for the divergence and it applies to any infrastructure.

> Actions are taken based on policies around [feedback](https://github.com/open-gitops/documents/blob/release-v1.0.0/GLOSSARY.md#feedback) from the system and previous reconciliation attempts, in order to reduce deviation over time.

AFAIU it is done by configuration management tools the same manner. They test the configuration, get the feedback from the system and try to execute action to reduce the drift.

In this case I see the difference in how feedback is defined:

> For example if the desired state requires more resources than exist in a system, the software agent may make attempts to add resources, to automatically rollback to a previous version, or to send alerts to human operators.

The "normal" configuration management tool is not capable of making "intelligent" decisions, such as rollback to a previous version or add more resources. It tries to get to the desired state in any case and if it is impossible, it fails.

> Don't get this confused. Principle \#3 doesn't mean pull vs push. It means the Git contents are pulled. But HOW it gets applied, doesn't matter.

Thank you for clarifying it! I fully agree with you, that is immense important, that the configuration is not only stored in Git, but also pulled from it before to be applied.

Should it be then something like:

The desired state declarations is pulled from the source.

Because it doesn't matter, **what** does it. 
btw "software agent" is not defined in the glossary.

**Reply by williamcaban** [reply]
I will like to expand in some concepts to help with some clarifications here.

If a tool or the integration of series of tools (e.g. automations tools drive by a pipelines on a continuous remediation loop) are setup in a way that they can follow principles 3 and 4, they pull + continuously reconcile, we can call that setup a "GitOps controller". Now, that alone, is not enough for calling it GitOps. After that, we must also make sure we have principle 2 in place. The intent behind that principle is very tied to principle 1 and the intent is what might have be lost. To illustrate the intent, lets say that my system is in "version 5", so, in that case, it is not only about "this is configuration v5" but to make sure in a future day, lets say 2yrs from now, if I want to run "version 5" of what I have today, that should result in an identical *system* to what I call today "version 5". 

Notice the use of the word [*system*](https://github.com/open-gitops/documents/blob/main/GLOSSARY.md#software-system) which as per glossary, includes the runtime environment, etc. Another way of seeing it following the example, it is like being able to do a "restore" procedure by using a "guaranteed rebuild" of "version 5" of the system. (It is not about an actual rebuild, but the outcome should be identical)

With this context, if you have a tool or combinations of tools where each one of the principles is followed in their whole context, then yes, you are doing GitOps.

_W
